### PR TITLE
JAMES-3693 Rate limiting should allow choosing precision

### PR DESCRIPTION
**Pull request by quantranhong1999**

### CODE REVIEW

**Review by chibenwa — COMMENTED** [review]
Conflicts with https://github.com/apache/james-project/pull/870 (please rebase)

we can make recision a parameter of `RateLimiterFactory::create` instead.

**Comment by chibenwa** [comment]
Can we please apply precision in all test classes implying mailets and rate limiting... Both unit and integration tests.

We would benefit from the additional stability...

File: server/mailet/rate-limiter/README.adoc [hunk 1]
```diff
@@ -40,6 +40,7 @@ Here are supported configuration parameters:
  - *exceededProcessor*: Processor to which emails whose rate is exceeded should be redirected to. Defaults to error.
  Use this to customize the behaviour upon exceeded rate.
  - *duration*: Duration during which the rate limiting shall be applied. Compulsory, must be a valid duration of at least one second. Supported units includes s (second), m (minute), h (hour), d (day).
+ - *precision*: Defines the time precision that will be used to approximate the sliding window. A lower duration increases precision. The precision must be greater than 1 second. Optional, default to duration means fixed window counters. Supported units includes s (second), m (minute), h (hour), d (day).
```
Comment (chibenwa) [inline server/mailet/rate-limiter/README.adoc:43]: A lower duration increases precision but requests more computing power.

File: server/mailet/rate-limiter/README.adoc [hunk 2]
```diff
@@ -84,6 +86,7 @@ this mailet twice within your `mailetcontainer.xml` file. Defaults to none.
 - *exceededProcessor*: Processor to which emails whose rate is exceeded should be redirected to. Defaults to error.
 Use this to customize the behaviour upon exceeded rate.
 - *duration*: Duration during which the rate limiting shall be applied. Compulsory, must be a valid duration of at least one second. Supported units includes s (second), m (minute), h (hour), d (day).
+- *precision*: Defines the time precision that will be used to approximate the sliding window. A lower duration increases precision. The precision must be greater than 1 second. Optional, default to duration means fixed window counters. Supported units includes s (second), m (minute), h (hour), d (day).
```
Comment (chibenwa) [inline server/mailet/rate-limiter/README.adoc:89]: A lower duration increases precision but requests more computing power.

Comment (chibenwa) [reply server/mailet/rate-limiter/README.adoc:89]: How about documenting precision for PerRcptRateLimiter?

File: server/mailet/rate-limiter/src/main/scala/org/apache/james/transport/mailets/GlobalRateLimit.scala [hunk 3]
```diff
@@ -66,6 +66,7 @@ case class GlobalRateLimiter(rateLimiter: Option[RateLimiter], keyPrefix: Option
  *    <li><b>exceededProcessor</b>: Processor to which emails whose rate is exceeded should be redirected to. Defaults to error.
  *    Use this to customize the behaviour upon exceeded rate.</li>
  *    <li><b>duration</b>: Duration during which the rate limiting shall be applied. Compulsory, must be a valid duration of at least one second. Supported units includes s (second), m (minute), h (hour), d (day).</li>
+ *    <li><b>precision</b>: Defines the time precision that will be used to approximate the sliding window. A lower duration increases precision. The precision must be greater than 1 second. Optional, default to duration means fixed window counters. Supported units includes s (second), m (minute), h (hour), d (day).</li>
```
Comment (chibenwa) [inline server/mailet/rate-limiter/src/main/scala/org/apache/james/transport/mailets/GlobalRateLimit.scala:69]: A lower duration increases precision but requests more computing power.

File: server/mailet/rate-limiter/src/main/scala/org/apache/james/transport/mailets/PerSenderRateLimit.scala [hunk 4]
```diff
@@ -66,6 +66,7 @@ case class SenderKey(keyPrefix: Option[KeyPrefix], entityType: EntityType, mailA
  *    <li><b>exceededProcessor</b>: Processor to which emails whose rate is exceeded should be redirected to. Defaults to error.
  *    Use this to customize the behaviour upon exceeded rate.</li>
  *    <li><b>duration</b>: Duration during which the rate limiting shall be applied. Compulsory, must be a valid duration of at least one second. Supported units includes s (second), m (minute), h (hour), d (day).</li>
+ *    <li><b>precision</b>: Defines the time precision that will be used to approximate the sliding window. A lower duration increases precision. The precision must be greater than 1 second. Optional, default to duration means fixed window counters. Supported units includes s (second), m (minute), h (hour), d (day).</li>
```
Comment (chibenwa) [inline server/mailet/rate-limiter/src/main/scala/org/apache/james/transport/mailets/PerSenderRateLimit.scala:69]: A lower duration increases precision but requests more computing power.

**Review by chibenwa — APPROVED** [review]

**Review by Arsnael — APPROVED** [review]

**Comment by quantranhong1999** [comment]
Can someone rerun the CI?

**Pull request merged**